crates/core/tedge_agent/src/software_manager/actor.rs

```diff
@@ -232,6 +241,47 @@ impl SoftwareManagerActor {
         Ok(())
     }
 
+    fn detect_self_update() -> Result<(), SoftwareManagerError> {
+        info!("Checking if tedge got self updated");
+        let current_running_version = env!("CARGO_PKG_VERSION");
+        info!("Current running version: {}", current_running_version);
+        if let Ok(executable_path) = std::env::current_exe() {
+            let agent_binary_path = executable_path.parent().unwrap().join(TEDGE_AGENT);
+            if let Ok(output) = Command::new(agent_binary_path).args(["--version"]).output() {
+                if output.status.success() {
+                    let version_output = String::from_utf8_lossy(&output.stdout);
+                    let version_output_split: Vec<&str> =
+                        version_output.split_whitespace().collect();
+                    if let ["tedge-agent", installed_binary_version] =
+                        version_output_split.as_slice()
+                    {
+                        info!("Installed binary version: {}", installed_binary_version);
+                        if current_running_version != *installed_binary_version {
+                            info!("Self update detected. Requesting shutdown...");
+                            return Err(SoftwareManagerError::NotRunningLatestVersion);
+                        }
+                    } else {
+                        return Err(
+                            anyhow!("Unexpected version output: {:?}", version_output).into()
+                        );
+                    }
+                } else {
+                    return Err(anyhow!(
+                        "Fetching version from installed binary failed with {}",
+                        String::from_utf8_lossy(&output.stderr)
+                    )
+                    .into());
+                }
+            } else {
+                return Err(anyhow!("Failed to fetch version of installed binary").into());
+            }
+        } else {
+            return Err(anyhow!("Failed to retrieve running executable path").into());
+        }
```

Comment:
suggestion: Rewrite using early return

IMO code nested in such a way is a bit hard to read, looking at any `return Err(...)` case, it's hard to track it to the actual condition which it covers.